cmd/compgen/cmd/common/common.go

```diff
@@ -0,0 +1,75 @@
+package cmd
+
+import (
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+)
+
+type TemplateData struct {
+	Name        string
+	PackageName string
+}
+
+// Executes `go mod tidy`
+func Tidy(path string) {
+	cmd := exec.Command("go", "mod", "tidy", "-e")
+	cmd.Dir = path
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		panic(string(output))
+	}
+}
+
+func Render(source fs.FS, destination string, data TemplateData) error {
+	// Iterate over all files/folders in 'source' and execute func() for entry found
+	return fs.WalkDir(source, "templates", func(path string, entry fs.DirEntry, err error) error {
+		// There was an error reading the fs. So we return the error to stop execution
+		if err != nil {
+			return err
+		}
+
+		// WalkDir starts with the root directory passed. So we skip the first iteration
+		if path == "templates" {
+			return nil
+		}
+
+		// We remove the top-level 'templates' directory
+		// E.g. "templates/main.go.tmpl" becomes "main.go.tmpl"
+		newFilename, err := filepath.Rel("templates", path)
+		if err != nil {
+			return err
+		}
+
+		newFilename = filepath.Join(destination, strings.TrimSuffix(newFilename, ".tmpl"))
+
+		// Copy directories
+		if entry.IsDir() {
+			return os.MkdirAll(newFilename, os.ModePerm)
+		}
+
+		// Note that 'path' is not unmodified
+		if strings.HasSuffix(path, ".tmpl") {
+			content, err := fs.ReadFile(source, path)
+			if err != nil {
+				return err
+			}
+
+			file, err := os.Create(newFilename)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			// The template is finally rendered and immediately written to file
+			tmpl := template.Must(template.New(newFilename).Parse(string(content)))
+			return tmpl.Execute(file, data)
+		}
+
+		// A non-directory, non-template file. Ignore it
```

Comment:
I don't follow this comment. I'm thinking it might be a hold over from previous commits and needs to be deleted. It doesn't make sense in the context of `return nil`

Reply:
Maybe it's just not worded well but that was intentionally included. It's meant to describe why we return nil. Here's a more verbose version

// The "entry" is not a directory and not a file with a name matching *.tmpl. So we skip this DirEntry.

Reply:
Still doesn't make sense. `nil` is returned because the function return has to return an error as a value, either the error or nil. That's why we have a return nil statement. The caller to `Render` would then check if an error was returned and act accordingly.

Reply:
Not quite. That `return nil` is embedded withing the dynamic [func()](https://pkg.go.dev/io/fs#WalkDirFunc) which is an argument to [fs.WalkDir()](https://pkg.go.dev/io/fs#WalkDir).

That dynamic func() is executed once for every entry found by fs.WalkDir. So the caller is fs.WalkDir, not Render(). Returning an error to fs.WalkDir stops the iteration over the directory entries. Another revision of the comment:

// The "entry" is not a directory and not a file with a name matching *.tmpl. We do nothing with these entries.

Reply:
I updated this comment. Feel free to give it another look

Reply:
yea, when we were on the call talking about testing, i noticed that upper return statement and has a suspicion I was wrong.